### PR TITLE
Fix panic, out of order chunks, and race warning during WAL replay

### PR DESCRIPTION
**Pull request by bboreham**
Closes https://github.com/prometheus/prometheus/issues/9846
Closes https://github.com/prometheus/prometheus/issues/9859

Possible fix for the race warning in #9847 

First we introduce `walSubsetProcessor`, which packages up the `processWALSamples()` function and its input and output channels, helping to clarify how these things relate.

Then extract a few more methods.

Finally we add a lock round each batch of samples, so the race detector can see that we never access series owned by the processor outside of a lock.

Although we have waited for the processor to finish, we still get a warning from the race detector because it doesn't know how the different parts relate.

### CODE REVIEW

**Review by codesome — COMMENTED** [review]
This PR as is LGTM. But I would like to also fix https://github.com/prometheus/prometheus/issues/9859 in this PR, it is a small change.

**Comment by pracucci** [comment]
> This PR as is LGTM. But I would like to also fix #9859 in this PR, it is a small change.

@bboreham I've added a test to reproduce issue #9859 in https://github.com/prometheus/prometheus/pull/9863. Do you mind picking it in this PR so we change the WAL replay to fix that too?

**Review by codesome — APPROVED** [review]

File: tsdb/head_wal.go [hunk 1]
```diff
@@ -378,20 +355,51 @@ func (h *Head) setMMappedChunks(mSeries *memSeries, mmc []*mmappedChunk) {
 		mSeries.mmMaxTime = mmc[len(mmc)-1].maxTime
 		h.updateMinMaxTime(mmc[0].minTime, mSeries.mmMaxTime)
 	}
+
+	// Any samples replayed till now would already be compacted. Resetting the head chunk.
+	mSeries.nextAt = 0
+	mSeries.headChunk = nil
```
Comment (pracucci) [inline tsdb/head_wal.go:361]: Note to reviewers: the key change to fix #9859 is that the head chunk is now reset after series creation while taking the new lock.

**Review by pracucci — APPROVED** [review]
Good job!

**Comment by alanprot** [comment]
Hi,

We update prometheus on Cortex and we noticed big difference on wall time replay (2x-4x) that seems related to this change:

Screen Shot 2022-05-09 at 8 27 13 PM

if we revert this change we cannot see this anymore.

is this expected?

**Comment by bboreham** [comment]
No, I would not expect that. 
This PR was mostly a refactoring, but we did add some locking. 
Please can you do some analysis where the program is spending the extra time.  Is it burning CPU, waiting on locks, etc.?  

Another angle might be to determine which commit within this PR makes the difference.

**Comment by alanprot** [comment]
Thanks for the quick response @bboreham 
I will do commit by commit and see what can I find..

**Comment by alanprot** [comment]
Seems that the first commit already causes the issue:

I guess that before we were returning early when the series was created and now we have to flush the input/output?


https://github.com/bboreham/prometheus/commit/1565c8803519ed87d002ae6c94466f3f62f7f114#diff-e55e448a1ce649167194e1ced19ab2009f422ad678011e333414b8de8e62d311L220

Screen Shot 2022-05-09 at 10 28 40 PM
 already increase the time:

**Comment by bboreham** [comment]
It's plausible. The "output" is a channel to re-use buffers; really it's waiting to make sure nothing is running in parallel.

Could you grab some gouroutine dumps while it runs, so we can see which line(s) are more popular to wait on?
E.g. 5x `/debug/pprof/goroutine?debug=1` should suggest something.

**Comment by alanprot** [comment]
The main difference in time is btw this 2 lines:

Current Version:

```
level=info ts=2022-05-10T17:05:55.020872066Z caller=head.go:542 org_id=user msg="Replaying WAL, this may take a while"
level=info ts=2022-05-10T17:07:05.268967653Z caller=head.go:578 org_id=user msg="WAL checkpoint loaded"
```

Reverted Version:
```
level=info ts=2022-05-10T17:12:04.844703353Z caller=head.go:542 org_id=user msg="Replaying WAL, this may take a while"
level=info ts=2022-05-10T17:12:09.31871618Z caller=head.go:578 org_id=user msg="WAL checkpoint loaded"
```

This is the goroutine dump while the process was hanging between those lines: https://gist.github.com/alanprot/217fff7d5ac2f3b53485bdebca4268fb

I do see the `waitUntilIdle` there.

```
1 @ 0x43b6f6 0x469c0e 0xe1859e 0xe15edf 0xe07b3a 0xdf7a5e 0xdf6c5b 0x1d271d0 0x1d295db 0xdc0067 0x46cfc1
#	0x469c0d	time.Sleep+0x12d														path-redated/lib/src/runtime/time.go:193
#	0xe1859d	path-redated/vendor/github.com/prometheus/prometheus/tsdb.(*walSubsetProcessor).waitUntilIdle+0x5d	path-redatedvendor/github.com/prometheus/prometheus/tsdb/head_wal.go:441
#	0xe15ede	path-redated/vendor/github.com/prometheus/prometheus/tsdb.(*Head).loadWAL+0x133e				path-redatedvendor/github.com/prometheus/prometheus/tsdb/head_wal.go:213
#	0xe07b39	path-redated/vendor/github.com/prometheus/prometheus/tsdb.(*Head).Init+0xf19				path-redatedvendor/github.com/prometheus/prometheus/tsdb/head.go:573
#	0xdf7a5d	path-redated/vendor/github.com/prometheus/prometheus/tsdb.open+0xb9d					path-redatedvendor/github.com/prometheus/prometheus/tsdb/db.go:750
#	0xdf6c5a	path-redated/vendor/github.com/prometheus/prometheus/tsdb.Open+0x9a					path-redatedvendor/github.com/prometheus/prometheus/tsdb/db.go:576
#	0x1d271cf	path-redated/vendor/github.com/cortexproject/cortex/pkg/ingester.(*Ingester).createTSDB+0x7ef		path-redatedvendor/github.com/cortexproject/cortex/pkg/ingester/ingester_v2.go:1611
#	0x1d295da	path-redated/vendor/github.com/cortexproject/cortex/pkg/ingester.(*Ingester).openExistingTSDB.func1+0x15a	path-redatedvendor/github.com/cortexproject/cortex/pkg/ingester/ingester_v2.go:1740
#	0xdc0066	path-redated/vendor/golang.org/x/sync/errgroup.(*Group).Go.func1+0x66
```

**Pull request merged**